src/main/java/org/openstreetmap/atlas/checks/validation/intersections/AtGradeSignPostCheck.java

```diff
@@ -0,0 +1,634 @@
+     * @return true if the edge is valid intersecting edge
+     */
+    private boolean isValidIntersectingEdge(final Edge edge)
+    {
+        return edge.isMasterEdge() && HighwayTag.highwayTag(edge).isPresent()
+                && this.highwayFilter.contains(edge.highwayTag().getTagValue());
+    }
+
+    /**
+     * Collect matching out edges and corresponding in edge in a map. Store the roundabout edges and
+     * non roundabout edges in separate maps.
+     *
+     * @param inEdges
+     *            List<Edge> inEdges
+     * @param outEdges
+     *            Set<Edge> outEdges
+     * @return Map<String, Map<AtlasEntity, Set<AtlasEntity>>> with Map of inEdge to outEdges for
+     *         roundabout and non roundabout edges
+     */
+    private Map<String, Map<AtlasEntity, Set<AtlasEntity>>> populateInEdgeToOutEdgeMaps(
+            final List<Edge> inEdges, final Set<Edge> outEdges)
+    {
+        final Map<AtlasEntity, Set<AtlasEntity>> inEdgeToOutEdgeMap = new HashMap<>();
+        final Map<AtlasEntity, Set<AtlasEntity>> roundAboutInEdgeToOutEdgeMap = new HashMap<>();
+        inEdges.forEach(inEdge ->
+        {
+            final Optional<HighwayTag> highwayTag = HighwayTag.highwayTag(inEdge);
+            if (highwayTag.isPresent()
+                    && this.connectedHighwayTypes.containsKey(highwayTag.get().getTagValue()))
+            {
+                // Filter out edges based on level and layer tags and valid highway types
+                final Set<AtlasEntity> filteredOutEdges = outEdges.stream()
+                        .filter(outEdge -> this.isMatchingOutEdge(inEdge, outEdge))
+                        .collect(Collectors.toSet());
+                // There should be at least 2 valid outEdges
+                if (filteredOutEdges.size() >= 2)
+                {
+                    final String inEdgeHighwayType = highwayTag.get().getTagValue();
+                    final List<String> validHighwayTypesOfOutEdge = this.connectedHighwayTypes
+                            .get(inEdgeHighwayType);
+                    final Set<AtlasEntity> filteredByHighways = filteredOutEdges.stream()
+                            .filter(atlasEntity ->
+                            {
+                                final Optional<HighwayTag> atlasEntityHighway = HighwayTag
+                                        .highwayTag(atlasEntity);
+                                return atlasEntityHighway.isPresent() && validHighwayTypesOfOutEdge
+                                        .contains(atlasEntityHighway.get().getTagValue());
+                            }).collect(Collectors.toSet());
+                    // If any of the edges is a roundabout, add it to roundabout map
+                    if (filteredByHighways.stream().anyMatch(JunctionTag::isRoundabout)
+                            || JunctionTag.isRoundabout(inEdge))
+                    {
+                        roundAboutInEdgeToOutEdgeMap.put(inEdge, filteredByHighways);
```

Comment:
Why don't we put only the edges in ```filteredByHighways``` that are explicitly roundabouts in ```roundAboutInEdgeToOutEdgeMap```? And the rest in ```inEdgeToOutEdgeMap```

Reply:
So if there is at least one roundabout edge, then the intersection should be in roundAboutInEdgeToOutEdgeMap. If one of the inEdge is an entrance to a roundabout and outEdge is exit to a roundabout, there will not be a roundabout tag to these edges and I do not want them to be added to inEdgeToOutEdgeMap. This way, I can be sure to assign roundabout instruction to the intersection of these edges.